File: src/dere_discord/message_handler.py
"""Message handler for Discord bot with clean typing indicator management."""

from __future__ import annotations

from typing import TYPE_CHECKING

import discord
from discord import AllowedMentions
from loguru import logger

if TYPE_CHECKING:
    from .agent import DiscordAgent
    from .persona import PersonaProfile


class TypingIndicatorContext:
    """Async context manager for Discord typing indicator with automatic cleanup."""

    def __init__(self, channel: discord.abc.Messageable):
        self.channel = channel
        self._typing_cm = None
        self._active = False

    async def __aenter__(self):
        self._typing_cm = self.channel.typing()
        await self._typing_cm.__aenter__()
        self._active = True
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        if self._active and self._typing_cm:
            try:
                await self._typing_cm.__aexit__(exc_type, exc_val, exc_tb)
            except Exception:
                # Suppress typing indicator errors to avoid noisy logging
                pass
            finally:
                self._active = False
        return False

    @property
    def is_active(self) -> bool:
        return self._active


class MessageHandlerCallbacks:
    """Encapsulates message handling callbacks to reduce nesting."""

    def __init__(
        self,
        message: discord.Message,
        build_embed_fn: callable,
    ):
        self.message = message
        self.build_embed_fn = build_embed_fn

    async def send_initial(self) -> None:
        """Send initial response (typing indicator already active)."""
        # Typing indicator already active; no placeholder message needed
        pass

    async def send_text_message(self, text: str, persona_profile: PersonaProfile) -> None:
        """Send text message to Discord channel."""
        content = text.strip() if text else ""
        if not content:
            return
        await self.message.channel.send(
            content,
            allowed_mentions=AllowedMentions.none(),
        )

    async def send_tool_summary(
        self, tool_events: list[str], persona_profile: PersonaProfile
    ) -> None:
        """Send tool summary embed to Discord channel."""
        if not tool_events:
            return
        embed = self.build_embed_fn(tool_events, persona_profile)
        await self.message.channel.send(
            embed=embed,
            allowed_mentions=AllowedMentions.none(),
        )


async def handle_discord_message(
    message: discord.Message,
    agent: DiscordAgent,
    guild_id: int | None,
    channel_id: int,
    user_id: int,
    content: str,
    build_embed_fn: callable,
) -> None:
    """Handle a Discord message with automatic typing indicator management.

    Args:
        message: Discord message object
        agent: Discord agent for handling the message
        guild_id: Guild ID if in a guild, None for DMs
        channel_id: Channel ID
        user_id: User ID
        content: Message content
        build_embed_fn: Function to build embed responses
    """
    callbacks = MessageHandlerCallbacks(message, build_embed_fn)

    async with TypingIndicatorContext(message.channel) as _:
        try:
            # Finalize callback that checks typing state
            async def finalize() -> None:
                # Typing indicator will be automatically cleaned up by context manager
                pass

            await agent.handle_message(
                guild_id=guild_id,
                channel_id=channel_id,
                user_id=user_id,
                content=content,
                send_initial=callbacks.send_initial,
                send_text_message=callbacks.send_text_message,
                send_tool_summary=callbacks.send_tool_summary,
                finalize=finalize,
            )
        except Exception as exc:
            logger.exception("Failed handling message in channel {}: {}", channel_id, exc)
            await message.channel.send("Sorry, something went wrong while contacting Claude.")

File: packages/discord/src/message-handler.ts
import type { EmbedBuilder, Message, TextBasedChannelFields } from "discord.js";

import type { DiscordAgent } from "./agent.js";
import type { PersonaProfile } from "./persona.js";

type BuildEmbedFn = (toolEvents: string[], personaProfile: PersonaProfile) => EmbedBuilder;

function isTextBasedChannel(channel: unknown): channel is TextBasedChannelFields {
  return Boolean(
    channel &&
      typeof (channel as { isTextBased?: () => boolean }).isTextBased === "function" &&
      (channel as { isTextBased: () => boolean }).isTextBased(),
  );
}

class TypingIndicator {
  private interval: ReturnType<typeof setInterval> | null = null;
  private active = false;

  async start(channel: { sendTyping: () => Promise<void> }): Promise<void> {
    if (this.active) {
      return;
    }
    this.active = true;

    const send = async () => {
      try {
        await channel.sendTyping();
      } catch {
        // Ignore typing errors to avoid noisy logs.
      }
    };

    await send();
    this.interval = setInterval(() => {
      void send();
    }, 8000);
  }

  stop(): void {
    if (this.interval) {
      clearInterval(this.interval);
      this.interval = null;
    }
    this.active = false;
  }
}

class MessageHandlerCallbacks {
  private readonly message: Message;
  private readonly buildEmbedFn: BuildEmbedFn;

  constructor(message: Message, buildEmbedFn: BuildEmbedFn) {
    this.message = message;
    this.buildEmbedFn = buildEmbedFn;
  }

  private getChannel(): TextBasedChannelFields | null {
    return isTextBasedChannel(this.message.channel) ? this.message.channel : null;
  }

  async sendInitial(): Promise<void> {
    // Typing indicator already active; no placeholder message needed.
  }

  async sendTextMessage(text: string, _personaProfile: PersonaProfile): Promise<void> {
    const content = text?.trim();
    if (!content) {
      return;
    }
    const channel = this.getChannel();
    if (!channel) {
      return;
    }
    await channel.send({
      content,
      allowedMentions: { parse: [] },
    });
  }

  async sendToolSummary(toolEvents: string[], personaProfile: PersonaProfile): Promise<void> {
    if (toolEvents.length === 0) {
      return;
    }
    const embed = this.buildEmbedFn(toolEvents, personaProfile);
    const channel = this.getChannel();
    if (!channel) {
      return;
    }
    await channel.send({
      embeds: [embed],
      allowedMentions: { parse: [] },
    });
  }
}

export async function handleDiscordMessage(args: {
  message: Message;
  agent: DiscordAgent;
  guildId: string | null;
  channelId: string;
  userId: string;
  content: string;
  buildEmbedFn: BuildEmbedFn;
}): Promise<void> {
  const callbacks = new MessageHandlerCallbacks(args.message, args.buildEmbedFn);
  const typing = new TypingIndicator();
  const textChannel = isTextBasedChannel(args.message.channel) ? args.message.channel : null;
  if (textChannel) {
    await typing.start(textChannel);
  }

  const finalize = async () => {
    typing.stop();
  };

  try {
    await args.agent.handleMessage({
      guildId: args.guildId,
      channelId: args.channelId,
      userId: args.userId,
      content: args.content,
      callbacks: {
        sendInitial: () => callbacks.sendInitial(),
        sendTextMessage: (text, profile) => callbacks.sendTextMessage(text, profile),
        sendToolSummary: (events, profile) => callbacks.sendToolSummary(events, profile),
        finalize,
      },
    });
  } catch (error) {
    console.error(`Failed handling message in channel ${args.channelId}: ${String(error)}`);
    if (textChannel) {
      await textChannel.send({
        content: "Sorry, something went wrong while contacting Claude.",
        allowedMentions: { parse: [] },
      });
    }
    typing.stop();
  }
}
